### PR TITLE
feat: add ability to add and parse empty namespaced shares

### PR DESCRIPTION
**Pull request by evan-forbes**
## Description

We need to be able to easily add "namespaced padding" in order to have non-interactive defaults. This PR adds that. Finally felt confident enough in the current non-interactive defaults implementation to mark this portion ready for review, as the rest depends on it.

Closes: #XXX

### CODE REVIEW

File: types/share_splitting.go [hunk 1]
```diff
@@ -47,8 +62,10 @@ func (msw *MessageShareWriter) Export() NamespacedShares {
 	return shares
 }
 
+// note: as an optimization we can probably get rid of this if we just add
+// checks each time we write.
 func (msw *MessageShareWriter) sortMsgs() {
-	sort.Slice(msw.shares, func(i, j int) bool {
+	sort.SliceStable(msw.shares, func(i, j int) bool {
```
Comment (evan-forbes) [inline types/share_splitting.go:68]: this probably could have been its own PR, as this is desirable even without non-interactive defaults, but it is specifically needed for padding so I thought I'd include it here

File: types/share_merging.go [hunk 2]
```diff
@@ -212,71 +212,65 @@ func parseMsgShares(shares [][]byte) ([]Message, error) {
...
+			newMessage = false
+		// this entire share contains a chunk of message that we need to save
+		case msgLen > dataLen:
+			currentMsg.Data = append(currentMsg.Data, shares[i][consts.NamespaceSize:]...)
+		// this share contains the last chunk of data needed to complete the
+		// message
+		case msgLen <= dataLen:
+			remaining := msgLen - len(currentMsg.Data) + consts.NamespaceSize
+			currentMsg.Data = append(currentMsg.Data, shares[i][consts.NamespaceSize:remaining]...)
+			saveLatestMessage()
```
Comment (evan-forbes) [inline types/share_merging.go:268]: probably could have gotten away without refactoring this func, but I feel like this much easier to grok

File: types/shares_test.go [hunk 3]
```diff
@@ -439,6 +438,26 @@ func Test_parseMsgShares(t *testing.T) {
...
+	}
+	msgs.SortMessages()
+	msgWr.Write(msgs.MessagesList[0])
+	msgWr.WriteNamespacedPaddedShares(4)
+	msgWr.Write(msgs.MessagesList[1])
+	msgWr.WriteNamespacedPaddedShares(10)
+	pmsgs, err := parseMsgShares(msgWr.Export().RawShares())
+	require.NoError(t, err)
+	require.Equal(t, msgs.MessagesList, pmsgs)
+}
```
Comment (evan-forbes) [inline types/shares_test.go:459]: note that most of the testing is occurring in preexisting tests since this is mainly a refactor. We will also be testing this much more in the app when we actually start using namespaced padding in an upcoming PR

**Review by liamsi — COMMENTED** [review]
Looks good to me. I think @adlerjohn should have a second look as he came up with the non-interactive defaults.

File: types/share_merging.go [hunk 4]
```diff
@@ -212,71 +212,65 @@ func parseMsgShares(shares [][]byte) ([]Message, error) {
...
 	}
-
-	// set the first nid and current share
-	nid := shares[0][:consts.NamespaceSize]
-	currentShare := shares[0][consts.NamespaceSize:]
-	// find and remove the msg len delimiter
-	currentShare, msgLen, err := ParseDelimiter(currentShare)
-	if err != nil {
-		return nil, err
+	// msgs returned
```
Comment (SweeXordious) [inline types/share_merging.go:215]: If I understand right, this function wasn't able to parse padding, now it is, right?
So, creating a block using the non-interactive defaults rules is already implemented somewhere?

File: types/share_merging.go [hunk 5]
```diff
@@ -212,71 +212,65 @@ func parseMsgShares(shares [][]byte) ([]Message, error) {
...
+	// current msg
+	currentMsg := Message{}
+	// the current share contains the start of a new message
+	newMessage := true
+	// the len in bytes of the current chunk of data that will eventually become
+	// a message. This is identical to len(currentMsg.Data) + consts.MsgShareSize
+	// but we cache it here for readability
+	dataLen := 0
+
+	saveLatestMessage := func() {
```
Comment (SweeXordious) [inline types/share_merging.go:228]: Probably the name `saveMessage` or `saveCurrentMessage` would be less misleading. Personally, I thought there will be multiple messages parsed and cached, but we will only save the latest one. While, if I understand right, we're reading only one message at a time.

File: types/shares_test.go [hunk 6]
```diff
@@ -439,6 +438,26 @@ func Test_parseMsgShares(t *testing.T) {
 	}
 }
 
+func TestParsePaddedMsg(t *testing.T) {
```
Comment (SweeXordious) [inline types/shares_test.go:441]: Do we have tests for non-interactive defaults somewhere? Testing all scenarios?

**Review by SweeXordious — COMMENTED** [review]
Implementation looks good. But I still  don't fully understand non interactive defaults.
Left some questions when you have time please

File: types/shares_test.go (hunk 6 above)
Comment (evan-forbes) [reply types/shares_test.go:441]: I'll list my other comment here
> note that most of the testing is occurring in preexisting tests since this is mainly a refactor. We will also be testing this much more in the app when we actually start using namespaced padding in an upcoming PR

but this _should_ (famous last words) be testing the scenarios we need as there fortunately not that many for padding. The only thing this implementation needs to do is to ignore padding, and in a flattened set a of shares, this is simple since messages with a length delimiter of 0 **must** be padding.

the most important thing to test with this implementation is that it is equivalent with the old implementation and that it catches the edge cases of decoding a square. It appears to be using the existing tests. We have a test to crudely fuzz this (before go 1.18's official fuzzing) which simply runs `TestDataFromSquare` (which round trip tests random data) on a loop for some period of time. I ran it for a few minutes straight with no failing tests, which gives me confidence in that portion.

File: types/share_merging.go (hunk 4 above)
Comment (evan-forbes) [reply types/share_merging.go:215]: > So, creating a block using the non-interactive defaults rules is already implemented somewhere?

While I have some very crude local experimental branches with pieces here and there that are not ready for the light of day, mostly yes. I did however feel comfortable enough to spin this out, especially since the most needed thing is just an easy way to add padding.

https://github.com/celestiaorg/celestia-specs/blob/master/src/rationale/message_block_layout.md#caveats

Comment (evan-forbes) [reply types/share_merging.go:215]: I'm also okay with not merging this until the app implementation is ready

Comment (SweeXordious) [reply types/share_merging.go:215]: It's good. It's better to merge PRs and not stockpile them. I guess merging is better :rocket:

File: types/shares_test.go (hunk 6 above)
Comment (SweeXordious) [reply types/shares_test.go:441]: Gonna further dig into this. Thanks a lot.

File: types/shares_test.go [hunk 7]
```diff
@@ -439,6 +438,26 @@ func Test_parseMsgShares(t *testing.T) {
...
+		MessagesList: []Message{
+			randomSmallMsg,
+			randomLargeMsg,
+		},
+	}
+	msgs.SortMessages()
+	msgWr.Write(msgs.MessagesList[0])
+	msgWr.WriteNamespacedPaddedShares(4)
+	msgWr.Write(msgs.MessagesList[1])
+	msgWr.WriteNamespacedPaddedShares(10)
```
Comment (rootulp) [inline types/shares_test.go:455]: [question for my understanding] would the previous implementation fail if the message share writer contained these namespaced padded shares? I recognize the entire test is new so we don't have a failing run with the previous `parseMsgShares` implementation

**Review by rootulp — APPROVED** [review]
I couldn't fully follow the implementation of `parseMsgShares` so I proposed a few optional readability nits but the complexity actually lies in the iteration logic. 

If I understand correctly, the new unit test `TestParsePaddedMsg` covers the empty namespaced shares case so approving to unblock.

File: types/shares_test.go (hunk 7 above)
Comment (evan-forbes) [reply types/shares_test.go:455]: yeah, the previous implementation assumes that if we find a share with message with a delimiter length of zero (messages are [prepended with a length delimiter](https://github.com/celestiaorg/celestia-core/blob/481fd21db7dfeb2d4312692ead148ccedd2951e7/types/block.go#L1168)) then the rest of the message are tailpadding.

File: types/share_merging.go [hunk 8]
```diff
@@ -212,71 +212,65 @@ func parseMsgShares(shares [][]byte) ([]Message, error) {
...
-
-	// set the first nid and current share
-	nid := shares[0][:consts.NamespaceSize]
-	currentShare := shares[0][consts.NamespaceSize:]
-	// find and remove the msg len delimiter
-	currentShare, msgLen, err := ParseDelimiter(currentShare)
-	if err != nil {
-		return nil, err
+	// msgs returned
+	msgs := []Message{}
```
[the remark this answers is not in the document]
Comment (adlerjohn) [reply types/share_merging.go:216]: Named return values are really hard to follow in any non-trivial function. Strongly against them.

File: types/share_merging.go [hunk 9]
```diff
@@ -212,71 +212,65 @@ func parseMsgShares(shares [][]byte) ([]Message, error) {
...
+		switch {
+		case newMessage:
+			nextMsgChunk, nextMsgLen, err := ParseDelimiter(shares[i][consts.NamespaceSize:])
+			if err != nil {
+				return nil, err
+			}
+			// the current share is namespaced padding so we ignore it
+			if nextMsgLen == 0 {
+				continue
+			}
```
Comment (adlerjohn) [inline types/share_merging.go:245]: This seems like it would work, since the bytes are zero and therefore the length delimiter is zero. But this also seems kind of brittle, since it's not really the length delimiter that's zero but all the bytes, specifically. Is there some way of exposing this information better, rather than relying on implementation details of the delimiter encoding?

File: types/share_merging.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (rootulp) [reply types/share_merging.go:216]: I'm curious on why? I think named return values increase clarity because they 
1. Remove the need to declare the variable in the function body. This results in one less line of code.
2. Draw attention to the value that will be returned in the function signature. I see this as a readability improvement, especially when the returned variable is modified multiple times in the function body (i.e. an aggregate value like `total` or `msgs` as is the case in this example).

Named return values do enable bare returns which I think are less clear than explicit returns and should be avoided. Fortunately, Go linter [revive](https://golangci-lint.run/usage/linters/#revive) warns on bare returns.

[the remark this answers is not in the document]
Comment (rootulp) [reply types/share_merging.go:216]: It looks like Go lang linter https://github.com/blizzy78/varnamelen has an option to configure named return values but given its not enabled in this repo and not enabled by default.

Also I realize named returns are a subjective code readability thing so I don't want to start a bikeshedding debate on it. Happy to resolve this comment and avoid named returns in this codebase if that's preferred.

[the remark this answers is not in the document]
Comment (adlerjohn) [reply types/share_merging.go:216]: Bare returns are one big issue. The other is un-obvious side effects. Assigning to a named return value is a side-effect to an otherwise pure function, and there's nothing in the syntax to indicate this explicitly to the reader. Named return values are basically global variables.

[the remark this answers is not in the document]
Comment (adlerjohn) [reply types/share_merging.go:216]: If we do decide to use them, then we should use them everywhere.

[the remark this answers is not in the document]
Comment (rootulp) [reply types/share_merging.go:216]: Spun out a bikeshed issue: https://github.com/celestiaorg/celestia-core/issues/831. Resolving this comment as a won't do until we reach agreement on a codebase convention.

**Comment by evan-forbes** [comment]
will be incorporating the feedback provided here into a different PR in the app, as this code needs to be moved to the app for non-interactive defaults

**Comment by evan-forbes** [comment]
closing in favor of https://github.com/celestiaorg/celestia-app/pull/637

please see this [commit](https://github.com/celestiaorg/celestia-app/pull/637/commits/2ca0ed57ca93b4d2855baad20b4018e18972a595) for the incorporation of the remaining feedback here :purple_heart:
